Add <template for> for declarative out-of-order streaming

foolip opened the pull request:
Add support for `<template for>`, where the `for` attribute value identifies a `<?marker>` or `<?start>` processing instruction to be replaced. This makes out-of-order streaming in the initial response possible. 

Fixes https://github.com/whatwg/html/issues/11542.

<!--
Thank you for contributing to the HTML Standard! Please describe the change you are making and complete the checklist below if your change is not editorial.

When you submit this PR, and each time you edit this comment (including checking a checkbox through the UI!), PR Preview will run and update it. As such make any edits in one go and only after PR Preview has run.

If you think your PR is ready to land, please double-check that the build is passing and the checklist is complete before pinging.
-->

- [x] At least two implementers are interested (and none opposed):
   * Chromium (@noamr)
   * Gecko (https://github.com/mozilla/standards-positions/issues/1369)
   * WebKit (@annevk, https://github.com/WebKit/standards-positions/issues/628)
- [x] [Tests](https://github.com/web-platform-tests/wpt) are written and can be reviewed and commented upon at:
   * https://wpt.fyi/results/html/dom/partial-updates/tentative?label=master&label=experimental&aligned&q=template-for
   * https://github.com/web-platform-tests/wpt/pull/61156
- [x] [Implementation bugs](https://github.com/whatwg/meta/blob/main/MAINTAINERS.md#handling-pull-requests) are filed:
   * Chromium: https://issues.chromium.org/issues/431374376
   * Gecko: https://bugzilla.mozilla.org/show_bug.cgi?id=2053472
   * WebKit: https://bugs.webkit.org/show_bug.cgi?id=318869
- [x] Corresponding [HTML AAM](https://w3c.github.io/html-aam/) & [ARIA in HTML](https://w3c.github.io/html-aria/) issues & PRs: N/A
- [x] [MDN issue](https://github.com/whatwg/meta/blob/main/MAINTAINERS.md#handling-pull-requests) is filed: https://github.com/mdn/mdn/issues/845
- [x] The top of this comment includes a [clear commit message](https://github.com/whatwg/meta/blob/main/COMMITTING.md) to use. <!-- If you created this PR from a single commit, Github copied its message. Otherwise, you need to add a commit message yourself. -->

(See [WHATWG Working Mode: Changes](https://whatwg.org/working-mode#changes) for more details.)


<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://whatpr.org/html/11818/indices.html" title="Last updated on Jul 17, 2026, 9:51 AM UTC (8b0598f)">/indices.html</a>  ( <a href="https://whatpr.org/html/11818/ade4fcd...8b0598f/indices.html" title="Last updated on Jul 17, 2026, 9:51 AM UTC (8b0598f)">diff</a> )
<a href="https://whatpr.org/html/11818/infrastructure.html" title="Last updated on Jul 17, 2026, 9:51 AM UTC (8b0598f)">/infrastructure.html</a>  ( <a href="https://whatpr.org/html/11818/ade4fcd...8b0598f/infrastructure.html" title="Last updated on Jul 17, 2026, 9:51 AM UTC (8b0598f)">diff</a> )
<a href="https://whatpr.org/html/11818/parsing.html" title="Last updated on Jul 17, 2026, 9:51 AM UTC (8b0598f)">/parsing.html</a>  ( <a href="https://whatpr.org/html/11818/ade4fcd...8b0598f/parsing.html" title="Last updated on Jul 17, 2026, 9:51 AM UTC (8b0598f)">diff</a> )
<a href="https://whatpr.org/html/11818/scripting.html" title="Last updated on Jul 17, 2026, 9:51 AM UTC (8b0598f)">/scripting.html</a>  ( <a href="https://whatpr.org/html/11818/ade4fcd...8b0598f/scripting.html" title="Last updated on Jul 17, 2026, 9:51 AM UTC (8b0598f)">diff</a> )
<a href="https://whatpr.org/html/11818/timers-and-user-prompts.html" title="Last updated on Jul 17, 2026, 9:51 AM UTC (8b0598f)">/timers-and-user-prompts.html</a>  ( <a href="https://whatpr.org/html/11818/ade4fcd...8b0598f/timers-and-user-prompts.html" title="Last updated on Jul 17, 2026, 9:51 AM UTC (8b0598f)">diff</a> )

noamr commented on source:
```diff
@@ -140208,11 +140280,26 @@ document.body.appendChild(text);
+       <li><p>TODO: what to we need to do so that elements are inserted into the template element
+       and redirected by our pre-insertion steps? We need to undo whatever it is that causes nodes
+       to be inserted into the DocumentFragment normally.</p></li>
+      </ol>
```
That's the "adjusted insertion location" or whatever it's called

foolip commented:
I've made some additional changes but things don't quite make sense yet. The direction I'm heading in is:

- When the parser encounters an `<template contentmethod>` element, it doesn't insert it.
- In [insert an element at the adjusted insertion location](https://html.spec.whatwg.org/#insert-an-element-at-the-adjusted-insertion-location), if we're about to insert an element with a `contentname` attribute into such a `<template>` element:
  - Find the target element among the descendents of the element that the `<template>` element was in. That target is kept as a bookkeeping slot, the tree traversal only happens once.
  - For `contentmethod=replace`, remove the target and actually insert. (For the following cases, the element isn't inserted.)
  - For `contentmethod=replace-children`, remove the children.
  - For `contentmethod=prepend`, save the current first element of the target.
  - For `contentmethod=append`, there's nothing to do here.
- In [appropriate place for inserting a node](https://html.spec.whatwg.org/#appropriate-place-for-inserting-a-node) further adjust the location based on `contentmethod` after the existing foster parenting adjustments. (This happens before the above step, but isn't relevant in that case.)
- In [insert an element at the adjusted insertion location](https://html.spec.whatwg.org/#insert-an-element-at-the-adjusted-insertion-location), if we're about to insert _into_ an element with a `contentname` attribute and the bookkeeping all checks out:
    - For `contentmethod=replace-children` and `contentmethod=append`, just append.
    - For `contentmethod=prepend`, use the saved first element of the target.

There are options for where to store the bookkeeping. I initially put it on the `<template>` element but now think it might be easier to follow if the bookkeeping goes on the element with the `contentname` attribute. An implementation might keep it as extra information in the stack of open elements.

noamr commented:
> Okay, should that state be sticky, or what happens if the nodes later realign so that the check passes?

Yea I think it errors the whole thing

noamr commented:
Missing pieces following the TPAC session:
- Applying patches inside the fragment parser (e.g. `innerHTML`) should be guarded in the same way as declarative shadow root, to avoid breaking assumptions made by userland sanitizers like DOMPurify.
- Find a way to notify in case of an error instead of failing silently. e.g. dispatch some error event on the document with a reference to the detached template element.
- Ensure that this works correctly when patching the child text nodes of scripts/styles, given that those elements have been "closed" before.

foolip commented:
I've given some thoughts to error handling for `contentmethod=prepend`. Here's an example of the problem:

```html
<!doctype html>
<body>
<div contentname=foo><span id=refnode>will be removed</span></div>

<template contentmethod=prepend>
  <div contentname=foo>
    <p>this element is inserted</p>
    <!-- the script removes the "content target first child" node -->
    <script>window.refnode = document.getElementById('refnode'); refnode.remove();</script>
    <p>reference node is gone, can this element be inserted?</p>
    <!-- put the reference node back -->
    <script>document.querySelector('[contentname=foo]').appendChild(refnode);</script>
    <p>is this OK because the reference node is back?</p>
  </div>
  <div contentname=foo>
    <p>is this fine because we saved a new reference node?</p>
  </div>
</template>
```

Options on what constitutes an error:

- Reference node is not a child of target
- Reference node has no parent (such that any parent would do)

Options on handling:

- Error is transient and if the error condition goes away, we keep inserting
- Error state is sticky to the `<div contentname=foo>`, skipping the rest of that patch
- Error state is sticky to the `<template contentmethod=prepend>`, skipping the rest of the current patch and any others in the template
- For sticky error states, it could be set either when the condition is broken (`refnode.remove()`) or when we discover that it is broken (when inserting the node after the script)

Options on reporting:

- Nothing, nodes are silently dropped, at most with devtools warnings
- Queue a task to fire a bubbling "contenterror" event at the template element's would-be parent. Could also be the would-be parent's root, more like the "unhandledrejection".
  - Either one event for every failed insertion or
  - one event for the first error or summarizing all errors.

It looks like the parser currently never fires events while it's still running, the only events are "DOMContentLoaded" and "load". Anything else that seems be fired by parsing is actually triggered by some other algorithm run as a side effect of what the parser does.

I don't think that [parse errors](https://html.spec.whatwg.org/#parse-errors) are a good fit either, because the error can't easily be expressed in terms of the input, but the shape of the DOM tree.

My thinking now is that we should instead guarantee that the nodes are inserted, that we don't discard node midstream. For `contentmethod=prepend` the model I think could work is maintaining an insertion point within the target element that is updated if children are removed.

noamr commented on source:
```diff
@@ -68178,6 +68122,122 @@ interface <dfn interface>HTMLTemplateElement</dfn> : <span>HTMLElement</span> {
+     <li><p>Let <var>nextNode</var> be <var>currentNode</var>'s
+     <span>next sibling</span>.</p></li>
+
+     <li><p><span data-x="concept-node-remove">Remove</span> <var>currentNode</var>.</p></li>
```
This can have impllications if the node is an iframe with a `pagehide`, e.g. it can synchronously add more siblings or change things about the document. It's perhaps OK but an alternative would be to accumulate all the nodes before the `<?end>` and then remove them in one go.

annevk replied:
How would you remove them all at once? That's not a primitive that we have. How do we account for changes here?

noamr replied:
Changes in there won't have any effect. You'd have accumulated a list of nodes in this step, and then when we see `<?end>` we remove all of those one by one. If there is a `pagehide` or whatnot for one of them it wouldn't be able to change that list.

annevk replied:
That is probably a good idea.

We probably still need to add guards though because concept-node-remove for instance asserts node's parent is non-null.

noamr replied:
Right, those should generally behave like `Node.prototype.remove()`

foolip replied:
In 218ede5d8f7494182726c9110574e85ef7e11ebd I changed this to collect the nodes and then remove them, similar to https://dom.spec.whatwg.org/#concept-node-replace-all. However, that uses the *suppressObservers* flag.

foolip replied:
Parent null check and note added in a04cb372212ef7f3f11cbb8f4856f8f698981670.

annevk commented on source:
```diff
@@ -67944,6 +67881,7 @@ interface <dfn interface>HTMLTemplateElement</dfn> : <span>HTMLElement</span> {
   [<span>HTMLConstructor</span>] constructor();
 
   readonly attribute <span>DocumentFragment</span> <span data-x="dom-template-content">content</span>;
+  [<span>CEReactions</span>, <span data-x="xattr-Reflect">Reflect</span>="<span data-x="attr-template-for">for</span>"] attribute DOMString <dfn attribute for="HTMLTemplateElement" data-x="dom-template-htmlFor">htmlFor</dfn>;
```
Why do we use `htmlFor` here? I thought that was no longer needed given changes to ECMAScript? We should use `for`.

hsivonen replied:
OTOH, do we really want the naming convention to be different for the same-named attribute on different HTML elements?

annevk replied:
I guess it depends on if we try to proceed with https://github.com/whatwg/html/issues/9379 for the other cases.

noamr replied:
I think at this point it would be preferable to expose it as `htmlFor` and have aliases for all of them when we decide on #9379 rather than end up with a mix and match of `for` and `htmlFor`.

annevk replied:
I guess I don't feel super strongly myself, but web developers really seem to dislike the prefix so I'm somewhat hesitant to spread it further.

noamr replied:
I wonder if @jakearchibald or @tunetheweb have a strong opinion about this. My opinion right now is that we should fix all of these together but I could be convinced that drawing the line here makes more sense.

tunetheweb replied:
No strong opinion myself but do agree that it seems weird to fix this one (and not support `htmlFor` at all?) while `htmlFor` is still the "standard" way IIUC?

noamr replied:
Suggesting to resolve this comment for now. We are likely to get mixed opinions from developers. Let's create aliases for this together with #9379.

noamr replied:
Discussed this with @jakearchibald on #9379 as well. There is rough consensus on the plan of starting with `htmlFor` and aliasing all of those attributes to `for` (+ `class`) when we resolve on that.

foolip replied:
@annevk if that plans sounds good to you, can I resolve this?

foolip replied:
I'll go ahead and resolve, this is the last remaining open comment.

annevk commented on source:
```diff
@@ -67959,6 +67897,11 @@ interface <dfn interface>HTMLTemplateElement</dfn> : <span>HTMLElement</span> {
+  <p>The <dfn element-attr for="template"><code data-x="attr-template-for">for</code></dfn> content
+  attribute identifies the <code>ProcessingInstruction</code> that marks the location to insert
+  content at (<code data-x="">&lt;?marker></code>) or the start of the range to replace
+  (<code data-x="">&lt;?start></code>).</p>
```
This doesn't tell me what a valid value is for this attribute.

foolip replied:
fixed in c16a6408f4ed86ab3f0c5dd0d1a39ca5706631c2, but it still uses `data-x=""` for some things that should probably be linked.

annevk commented on source:
```diff
@@ -68231,6 +68293,62 @@ interface <dfn interface>HTMLTemplateElement</dfn> : <span>HTMLElement</span> {
 
 
+  <div class="example">
+
```
No newline here.

annevk commented on source:
```diff
@@ -153166,6 +153167,11 @@ interface <dfn interface>External</dfn> {
+     <th> <code data-x="">for</code>
+     <td> <code data-x="attr-template-for">template</code>
+     <td> Updates existing content
+     <td> Text
```
It looks like we're missing an update to the `template` element index entry.

foolip replied:
Added in 99241f909d57f9b03776e4e16e9af44c29c2016d.

annevk commented on source:
```diff
@@ -68178,6 +68122,124 @@ interface <dfn interface>HTMLTemplateElement</dfn> : <span>HTMLElement</span> {
+     <li><p>If <var>descendant</var> is not a <code>ProcessingInstruction</code> node, then
+     <span>continue</span>.</p></li>
+
+     <li><p>If <var>descendant</var>'s <code data-x="">name</code> attribute's value is not
```
Are we going to call these attributes or pseudo-attributes? Can we link this to a thing in DOM? And aren't they a map in DOM? We should use map syntax, no?

foolip replied:
I don't have a strong opinion about the naming, but think that since we're making them official it's not _wrong_ to call them plain attributes.

Something should be linked, yes. Rather than exposing the "attribute map" directly, how about we export algorithms like https://dom.spec.whatwg.org/#concept-element-attributes-get-by-name?

annevk replied:
That sounds fine

foolip replied:
Fixed in 32e6a437f2867f72069c8da87217500b472f170a.

annevk commented on source:
```diff
@@ -68178,6 +68122,124 @@ interface <dfn interface>HTMLTemplateElement</dfn> : <span>HTMLElement</span> {
+      <ol>
+       <li><p>Let <var>start</var> be <var>descendant</var>.</p></li>
+
+       <li><p>Let <var>nesting level</var> be 0.</p></li>
```
nestingLevel

annevk commented on source:
```diff
@@ -68178,6 +68122,124 @@ interface <dfn interface>HTMLTemplateElement</dfn> : <span>HTMLElement</span> {
+
+        <ol>
+         <li><p>If <var>sibling</var> is not a <code>ProcessingInstruction</code> node, then
+         <span>continue</span>.</p></li>
```
So now we are doing nested traversals. I'm not sure that actually works as the descendant iterator will also just pick the next one. Presumably we want to skip over the descendants we are handling here?

noamr replied:
This is not exactly "nested traversal". We reach the first matching PI and then sibling-traverse it, but after that we're done.
I double-checked this algorithm and I think it does the right thing... Can you be more specific if there is an issue here?

foolip replied:
I changed this bit today to fix an infinite loop. I think it's not broken any longer, but please comment again if I'm wrong.

annevk commented on source:
```diff
@@ -68178,6 +68122,124 @@ interface <dfn interface>HTMLTemplateElement</dfn> : <span>HTMLElement</span> {
+           <li><p>If <var>nesting level</var> is 0, then return (<var>start</var>,
+           <var>sibling</var>).</p></li>
+
+           <li><p>Otherwise, decrement <var>nesting level</var> by one.</p></li>
```
No need for otherwise with early returns.

annevk commented on source:
```diff
@@ -140570,6 +140510,67 @@ document.body.appendChild(text);
+        <p class="note">This is to support patching inside plain <code>template</code> elements or
+        those using <code data-x="attr-template-shadowrootmode">shadowrootmode</code>. Nested
+        patching is not supported, since in this case <var>scope</var>'s <span>template
+        contents</span> will be empty and <span>prepare content patching</span> will fail.</p>
```
"Plain" template contents are a `DocumentFragment`. This means you have a bunch of type errors as you exclude that earlier.

noamr replied:
Where is it excluded? `scope` can be a `DocumentFragment` even before this (see line 61835).

foolip replied:
Can you elaborate, are you saying that this note is incorrect, or that _scope_ isn't guaranteed to be an `Element` or `DocumentFragment` like I think it is?

annevk replied:
```
<p>Each <code>template</code> element has an associated <code>Element</code>-or-<code>ShadowRoot</code>-or-null <dfn>insertion target</dfn>, which is initially null.</p>
```

foolip replied:
Fixed in 9cd376fbfde2f115ad8865f677070551ad0f2fad.

annevk commented on source:
```diff
@@ -140570,6 +140510,67 @@ document.body.appendChild(text);
+        <p class="note">This is to support patching <code>head</code> with a <code>template</code>
+        inside <code>body</code>.</p>
+       </li>
+
```
We should have an assert here that scope is an Element or DocumentFragment.

Though also, can "the body element" be made to return null above with script?

noamr replied:
If the document doesn't have a body then the condition after "otherwise" is never true. I think we can assert that the scope is an element or document fragment.

foolip replied:
In ac1f4783e900c9e8e28e87ef466ed9642d3cca13 I've added the assert to the start of "prepare content patching", which is after these steps adjusting *scope*. I could hoist it to after these steps that adjust *scope*, let me know if you'd prefer that. Resolving for now.

foolip commented on source:
```diff
@@ -68231,6 +68322,61 @@ interface <dfn interface>HTMLTemplateElement</dfn> : <span>HTMLElement</span> {
   </div>
 
 
+  <div class="example">
```
Add a second example that uses different PI syntax to really resolve the discussion in https://github.com/whatwg/html/pull/12118#discussion_r3412823140.

foolip replied:
Fixed in 0040c46045e1cc1874a1f0012fe5de0e1d42af57.

foolip committed:
Add <template for> for declarative out-of-order streaming

Add support for <template for>, where the for attribute value
identifies a <?marker> or <?start> processing instruction to be
replaced.

This makes out-of-order streaming in the initial response possible.

Fixes https://github.com/whatwg/html/issues/11542

noamr commented on source:
```diff
@@ -145495,6 +145769,68 @@ document.body.appendChild(text);
+
+       <li>
+        <p>If <var>scope</var> is a <code>template</code> element, then set <var>scope</var> to
+        <var>scope</var>'s <span>template contents</span>.</p>
```
```suggestion
        <p>Let <var>scope</var> be the <span>adjusted insertion location</span>.</p>

        <p class="note">In the <span>fragment case</span>, <var>scope</var> can be the
        <code>DocumentFragment</code> into which nodes are being inserted to.</p>
       </li>

       <li>
        <p>If <var>scope</var> is a <code>template</code> element, then set <var>scope</var> to
        <var>scope</var>'s <span>template contents</span>.</p>
```

foolip committed:
Apply Noam's suggestion

foolip committed:
Address Anne's feedback

foolip committed:
Add back missing "insert an HTML element"

foolip committed:
Scope is a node, not a location in a node

foolip committed:
grammar

annevk committed:
grammar nits

annevk committed:
handle incorrect nesting inside template more consistently

annevk committed:
one more nit

annevk reviewed (approved):
I made a bunch more tweaks and also fixed a substantive issue with unclosed elements. My changes will need to be reviewed. As well as the tests I added.

I'm not entirely sure around the editorial convention of `<?blah>` for processing instructions, but I guess it's okay for now.

It would be ideal if @hsivonen or @zcorpan also reviewed this change given their parser knowledge.

noamr commented:
The tweaks are fine with me... @hsivonen @zcorpan mind taking a look so that we can land this?

hsivonen reviewed (approved):
Looks like GitHub can't give me a diff from what I've reviewed previously, but let's land this. Sorry about the delay again.

noamr commented:
@annevk I see that you've merged the WPTs. Does this mean that we can merge this one as well or are you still looking?
